### ISSUE TITLE
Hooks seem to be executing on macro-expanded code

### ISSUE THREAD

**Issue by pjsmith** [clj-kondo/clj-kondo#1907]
**Version**
`clj-kondo v2022.11.02`

**Platform**
Installed via `brew` on MacOS -- `brew install clj-kondo`. I think that's the native version

**Editor**
N/A

**Problem**

I've got a clj-kondo hook which tells me when I'm threading a value through only one form: catching issues like `(-> foo :bar)`, which could be expressed as `(:bar foo)` instead.

When I assign this hook via `[:hooks :analyze-call]` in my config.edn, I get correct warnings, but I also get some false positives. The false positives look like forms which are produced during macro expansion of the source files; I expected hooks to be running on _verbatim_ source code, not macro-expanded code. If a macro expands and produces a nasty form, it doesn't seem like something I want clj-kondo to worry about.

I'm running clj-kondo at the command line, not using an editor integration.

(Note: I raised this [on Stack Overflow](https://stackoverflow.com/questions/74637836/can-i-execute-a-clj-kondo-hook-before-any-macro-expansions-occur/74642450?noredirect=1#comment131832989_74642450) and @borkdude suggested I bring it here)

**Repro**

The hook doing the analysis (looking for forms like `(-> foo :bar)`)
```
;; ./.clj-kondo/peter/analyzers.clj

(ns peter.analyzers
  (:require
   [clj-kondo.hooks-api :as api]))

(defn superfluous-arrow
  [{:keys [node]}]
  (let [[arrow _data & forms] (:children node)]
    (when (= 1 (count forms))
      (api/reg-finding!
       (assoc (meta node)
              :message (format "%s: no need to thread a single form - %s (meta %s)" arrow node (meta node))
              :type :peter.analyzers/superfluous-arrow)))))
```

This file causes problems when analyzed with the hook:

```
(ns repro)

(cond->> "fuzz"
  (= "fizz") (pr "fizz")
  (= "buzz") (pr "buzz"))
```

I get this output on running `clj-kondo --lint repro.clj --config ./.clj-kondo/config.edn` (config as described below)

```
repro.clj::: warn: clojure.core/->>: no need to thread a single form - (clojure.core/->> G__3 (pr "fizz"))
repro.clj::: warn: clojure.core/->>: no need to thread a single form - (clojure.core/->> G__3 (pr "buzz"))
repro.clj:4:3: warning: Single operand use of clojure.core/= is always true
repro.clj:5:3: warning: Single operand use of clojure.core/= is always true
linting took 10ms, errors: 0, warnings: 2
```
The first two warnings seem to be analyzing macro-expanded forms: `(clojure.core/->> G__3 (pr "fizz"))` and `(clojure.core/->> G__3 (pr "buzz"))`

**Config**

I've reproduced this with a config having nothing in it other than my hooks:

```
{:linters {:peter.analyzers/superfluous-arrow {:level :warn}}
 :hooks {:analyze-call {clojure.core/-> peter.analyzers/superfluous-arrow
                        clojure.core/->> peter.analyzers/superfluous-arrow}}}
```

**Expected behavior**

I would expect hooks to analyze source code _verbatim_, before macro expansions are applied. If a macro expands to something that offends my hook, I don't want a warning about it.

**Comment by borkdude** [comment]
@pjsmith I added a `generated-node?` predicate in the hooks API that should only return `true` when the node was generated. Can you try this?

**Comment by borkdude** [comment]
Your hook would then be changed to:

```
(when-not (generated-node? node) ...)
```

**Comment by pjsmith** [comment]
That works perfectly, thanks @borkdude -- I'll update my StackOverflow answer for future searchers, and if I get a chance I'll open a PR to mention this in the hooks documentation.
